src/Equinox.MessageDb/MessageDb.fs

```diff
@@ -0,0 +1,435 @@
+                "Resync", Stats.LogSink.Resync ]
+            let logActivity name count lat =
+                log.Information("{name}: {count:n0} requests; Average latency: {lat:n0}ms",
+                    name, count, (if count = 0L then Double.NaN else float lat/float count))
+            let mutable rows, totalCount, totalMs = 0, 0L, 0L
+            for name, stat in stats do
+                if stat.count <> 0L then
+                    totalCount <- totalCount + stat.count
+                    totalMs <- totalMs + stat.ms
+                    logActivity name stat.count stat.ms
+                    rows <- rows + 1
+            // Yes, there's a minor race here between the use of the values and the reset
+            let duration = Stats.LogSink.Restart()
+            if rows > 1 then logActivity "TOTAL" totalCount totalMs
+            let measures : (string * (TimeSpan -> float)) list = [ "s", fun x -> x.TotalSeconds(*; "m", fun x -> x.TotalMinutes; "h", fun x -> x.TotalHours*) ]
+            let logPeriodicRate name count = log.Information("rp{name} {count:n0}", name, count)
+            for uom, f in measures do let d = f duration in if d <> 0. then logPeriodicRate uom (float totalCount/d |> int64)
+
+
+module private Write =
+    let private writeEventsAsync (log : ILogger) (conn : MessageDbClient) (streamName : string) (version : int64) (events : IEventData<EventBody> array)
+        : Async<MdbSyncResult> = async {
+            let! wr = conn.WriteMessages(streamName, events, version, Async.DefaultCancellationToken) |> Async.AwaitTaskCorrect
+            return
+              match wr with
+              | MdbSyncResult.ConflictUnknown ->
+                  log.Information("MdbTrySync WrongExpectedVersionException writing {EventTypes}, expected {ExpectedVersion}",
+                                  [| for x in events -> x.EventType |], version)
+                  wr
+              | _ -> wr }
+
+    let inline len (bytes: EventBody) = bytes.Length
+    let eventDataBytes events =
+        let eventDataLen (x : IEventData<EventBody>) = len x.Data + len x.Meta
+        events |> Array.sumBy eventDataLen
+    let private writeEventsLogged (conn : MessageDbClient) (streamName : string) (version : int64) (events : IEventData<EventBody> array) (log : ILogger)
+        : Async<MdbSyncResult> = async {
+        let log = if (not << log.IsEnabled) Events.LogEventLevel.Debug then log else log |> Log.propEventData "Json" events
+        let bytes, count = eventDataBytes events, events.Length
+        let log = log |> Log.prop "bytes" bytes
+        let writeLog = log |> Log.prop "stream" streamName |> Log.prop "expectedVersion" version |> Log.prop "count" count
+        let! t, result = writeEventsAsync writeLog conn streamName version events |> Stopwatch.Time
+        let reqMetric : Log.Measurement = { stream = streamName; interval = t; bytes = bytes; count = count}
+        let resultLog, evt =
+            match result, reqMetric with
```

Comment:
not sure why `reqMetric` is tupled into this match - replace `m` with inlined?

Reply:
Yep, that's better indeed (think this is leftover from my hasty copy from the SSS impl